keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,498 @@
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This enhancement introduces workload preemption, a mechanism to suspend
+workloads when:
+- ClusterQueues under their minimum quota need the resources that are currently
+  lent to other ClusterQueues in the cohort.
+- Within a Clusterqueue, there are running Workloads with lower priority than
+  a pending Workload.
+
+API fields in the Workload, LocalQueue and ClusterQueue objects control
+different preemption policies.
+
+## Motivation
+
+When ClusterQueues under their minimum quota lend resources, they should
+be able to recover those resources fast, to be able to admit Workloads
+when there are sudden spikes. Similarly, the ClusterQueue should be able to
+recover quota from low priority workloads that are currently running.
+
+Currently, the only mechanism to recover those resources is to wait for
+Workloads to finish, which is generally unbounded.
+
+### Goals
+
+- Preempt Workloads from ClusterQueues borrowing resources when other
+  ClusterQueues in the cohort, under their minimum quota, need the resources.
+- Preempt Workloads within a ClusterQueue when a high priority Workload doesn't
+  fit in the available quota, independently of borrowed quota.
+- Introduce API fields in Workload, LocalQueue and ClusterQueue to inform
+  which Workloads can be preempted and in which order.
+
+### Non-Goals
+
+- Graceful termination of Workloads is left to the workload pods to implement.
+- Tracking usage by workloads that take arbitrary time to be suspended. See
+  [Workload preemption doesn't imply immediate Pod termination](#workload-preemption-doesnt-imply-immediate-pod-termination) to learn more.
+  For example, the integration with Job uses the [suspend field](https://kubernetes.io/docs/concepts/workloads/controllers/job/#suspending-a-job).
+- Partial workload suspension is not supported.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+This enhancement proposes the introduction of the following APIs to influence
+Workload preemption:
+- A field in the Workload to describe the cost of preempting a Workload at a
+  given time.
+- A field in the LocalQueue to set Workload defaults.
+- A field in the ClusterQueue to influence how long to wait before starting
+  preemption.
+
+And an algorithm for selecting a set of Workloads using borrowed quota to be
+preempted when a Workload in a ClusterQueue is under the min quota.
```

Comment:
What happens to preempted workloads, do they re-enter the queue or finish as failed?

Reply:
Added as a knob